File: backend/schemas/risk.py
"""Pydantic schemas for risk event workflow APIs."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class RiskMeasureBase(BaseModel):
    step: str = Field(description="step3_execute_measures or step4_follow_up")
    measure_type: str
    description: str
    owner: Optional[str] = None
    status: str = Field(default="open")
    result: Optional[str] = None
    extra: Dict[str, Any] = Field(default_factory=dict)


class RiskMeasureOut(RiskMeasureBase):
    id: str
    created_at: datetime
    updated_at: datetime


class RiskDecisionCreate(BaseModel):
    summary: str
    decision: str
    committee_members: List[Dict[str, Any]] = Field(default_factory=list)
    attachments: List[Dict[str, Any]] = Field(default_factory=list)


class RiskDecisionOut(RiskDecisionCreate):
    id: str
    created_at: datetime
    updated_at: datetime


class RiskEventCreate(BaseModel):
    title: str
    event_type: str
    severity: str
    description: Optional[str] = None
    source_role: Optional[str] = None
    reporter_id: Optional[str] = None
    related_account_id: Optional[str] = None
    related_strategy_id: Optional[str] = None
    metadata: Dict[str, Any] = Field(default_factory=dict)


class RiskEventOut(BaseModel):
    id: str
    title: str
    event_type: str
    severity: str
    status: str
    step: str
    description: Optional[str] = None
    source_role: Optional[str] = None
    reporter_id: Optional[str] = None
    related_account_id: Optional[str] = None
    related_strategy_id: Optional[str] = None
    detected_at: Optional[datetime] = None
    resolved_at: Optional[datetime] = None
    metadata: Dict[str, Any] = Field(default_factory=dict)
    created_at: datetime
    updated_at: datetime


class RiskEventDetail(RiskEventOut):
    decisions: List[RiskDecisionOut] = Field(default_factory=list)
    measures: List[RiskMeasureOut] = Field(default_factory=list)


class RiskTransitionRequest(BaseModel):
    to_step: str
    operator: str
    reason: str

File: frontend/src/services/risk.ts
import { http } from './http'

export type RiskEvent = {
  id: string
  title: string
  event_type: string
  severity: string
  status: string
  step: string
  description?: string | null
  source_role?: string | null
  reporter_id?: string | null
  related_account_id?: string | null
  related_strategy_id?: string | null
  detected_at?: string | null
  resolved_at?: string | null
  metadata?: Record<string, any>
  created_at: string
  updated_at: string
}

export type RiskDecision = {
  id: string
  summary: string
  decision: string
  committee_members: Array<Record<string, any>>
  attachments: Array<Record<string, any>>
  created_at: string
  updated_at: string
}

export type RiskMeasure = {
  id: string
  step: string
  measure_type: string
  description: string
  owner?: string | null
  status: string
  result?: string | null
  extra: Record<string, any>
  created_at: string
  updated_at: string
}

export type RiskEventDetail = RiskEvent & {
  decisions: RiskDecision[]
  measures: RiskMeasure[]
}

export async function listRiskEvents(status?: string): Promise<RiskEvent[]> {
  const res = await http.get('/v1/risk/events', { params: status ? { status } : {} })
  return res.data
}

export async function createRiskEvent(payload: {
  title: string
  event_type: string
  severity: string
  description?: string
  source_role?: string
  reporter_id?: string
  related_account_id?: string
  related_strategy_id?: string
  metadata?: Record<string, any>
}): Promise<RiskEvent> {
  const res = await http.post('/v1/risk/events', payload)
  return res.data
}

export async function getRiskEventDetail(eventId: string): Promise<RiskEventDetail> {
  const res = await http.get(`/v1/risk/events/${eventId}`)
  return res.data
}

export async function addRiskDecision(eventId: string, payload: Omit<RiskDecision, 'id' | 'created_at' | 'updated_at'>) {
  const res = await http.post(`/v1/risk/events/${eventId}/decision`, payload)
  return res.data as RiskDecision
}

export async function addRiskMeasure(eventId: string, payload: Omit<RiskMeasure, 'id' | 'created_at' | 'updated_at'>) {
  const res = await http.post(`/v1/risk/events/${eventId}/measures`, payload)
  return res.data as RiskMeasure
}

export async function transitionRiskEvent(eventId: string, payload: { to_step: string; operator: string; reason: string }) {
  const res = await http.post(`/v1/risk/events/${eventId}/transition`, payload)
  return res.data as RiskEvent
}

